Docs/types.md

```diff
@@ -0,0 +1,645 @@
+
+// GoogleIdentityProvider provides identities for users authenticating using Google credentials
+type GoogleIdentityProvider struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// ClientID is the oauth client ID
+	ClientID string `json:"clientID"`
+	// ClientSecret is the oauth client secret
+	ClientSecret StringSource `json:"clientSecret"`
+
+	// HostedDomain is the optional Google App domain (e.g. "mycompany.com") to restrict logins to
+	HostedDomain string `json:"hostedDomain"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// OpenIDIdentityProvider provides identities for users authenticating using OpenID credentials
+type OpenIDIdentityProvider struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CA is the optional trusted certificate authority bundle to use when making requests to the server
+	// If empty, the default system roots are used
+	CA string `json:"ca"`
+
+	// ClientID is the oauth client ID
+	ClientID string `json:"clientID"`
+	// ClientSecret is the oauth client secret
+	ClientSecret StringSource `json:"clientSecret"`
+
+	// ExtraScopes are any scopes to request in addition to the standard "openid" scope.
+	ExtraScopes []string `json:"extraScopes"`
+
+	// ExtraAuthorizeParameters are any custom parameters to add to the authorize request.
+	ExtraAuthorizeParameters map[string]string `json:"extraAuthorizeParameters"`
+
+	// URLs to use to authenticate
+	URLs OpenIDURLs `json:"urls"`
+
+	// Claims mappings
+	Claims OpenIDClaims `json:"claims"`
+}
+
+// OpenIDURLs are URLs to use when authenticating with an OpenID identity provider
+type OpenIDURLs struct {
+	// Authorize is the oauth authorization URL
+	Authorize string `json:"authorize"`
+	// Token is the oauth token granting URL
+	Token string `json:"token"`
+	// UserInfo is the optional userinfo URL.
+	// If present, a granted access_token is used to request claims
+	// If empty, a granted id_token is parsed for claims
+	UserInfo string `json:"userInfo"`
+}
+
+// OpenIDClaims contains a list of OpenID claims to use when authenticating with an OpenID identity provider
+type OpenIDClaims struct {
+	// ID is the list of claims whose values should be used as the user ID. Required.
+	// OpenID standard identity claim is "sub"
+	ID []string `json:"id"`
+	// PreferredUsername is the list of claims whose values should be used as the preferred username.
+	// If unspecified, the preferred username is determined from the value of the id claim
+	PreferredUsername []string `json:"preferredUsername"`
+	// Name is the list of claims whose values should be used as the display name. Optional.
+	// If unspecified, no display name is set for the identity
+	Name []string `json:"name"`
+	// Email is the list of claims whose values should be used as the email address. Optional.
+	// If unspecified, no email is set for the identity
+	Email []string `json:"email"`
+}
+
+// GrantConfig holds the necessary configuration options for grant handlers
+type GrantConfig struct {
```

Comment:
unsure if these make sense. probably can move Method to default to `prompt` and let users override per provider and don't need `ServiceAccountMethod`